bokeh/models/ranges.py

```diff
@@ -119,6 +142,16 @@ class FactorRange(Range):
     this categorical range.
     """)
 
+    bound_lower = Either(String, Int, help="""
+    The minimum value that the range is allowed to go to - typically used to prevent
+    the user from panning/zooming/etc away from the data.
+    """)
+
+    bound_upper = Either(String, Int, help="""
+    The max value that the range is allowed to go to - typically used to prevent
+    the user from panning/zooming/etc away from the data.
+    """)
```

Comment:
I agree with Sarah that allowing panning beyond the data range is undesirable behavior in most cases.  However, it should be noted that we oftentimes will overlay multiple datasets and plots onto the same plot area, in which case we will want to be able to pan/zoom beyond the boundary of any individual plot.

Bryan is also correct that having "lower" and "upper" as bounds implies an ordering *at the data level*, which does not necessarily exist for sets of factors.  Stated differently, even with these parameters available, without also providing the user some understanding of how their unordered data will be sorted, then they can't reason about what it means for two random elements from the set to be "min" and "max".

However, it should be noted here that at some point we *do* splat things onto the screen, and there is a draw order imposed at that time (or by that time).  In that context, then, it does make sense to refer to the leftmost/rightmost/bottommost/topmost data element.

I like the suggestion that really what we're trying to express is "bounds policy", and "min"/"max" is a shorthand for a particular "bounds policy" or "fit policy", which is "Limit to tight fit at a precise literal values", with a tuple expressing min and max.  For unordered sets, one could imagine a bounds policy that is just "Limit to tight data fit", or "Limit to tight data fit, with a padding of 10 pixels".  Alternatively, one can imagine bounds policy that is an offset: "Limit upper bound to maximum data data; lower bound defaults to offset of 12.3 data units".

Where is that #MidnightIdeation tag... Oh, found it.